docs/unification.md

```diff
@@ -0,0 +1,344 @@
+  constructors.
+
+If one term is a variable and the other is a function-like pattern (but not a
+variable), consider the following cases:
+
+* If the variable is free, bind it to the pattern.
+* If the variable is bound to a variable, bind the variable it is bound to to
+  the pattern.
+* If the variable is bound to a constructor, decompose with the constructor and
+  the pattern.
+
+Constructors
+------------
+
+If the first term is an injective application and the second is the same
+symbol, decompose with each pair of children of each pattern.  If both terms
+are different constructors or overloaded symbols, fail.
+
+Injections
+----------
+
+If both terms are `inj`, and you are unifying `inj{s1, s}(p1)` and
+`inj{s2, s}(p2)`:
+
+* If s1 == s2, decompose with `p1` and `p2`.
+* If s1 < s2, decompose with `inj{s1, s2}(p1)` and `p2`.
+* If s2 < s1, symmetric of above
+* otherwise, fail
+
+If one term is `inj` and the other is a constructor, fail.
+
+Overloading
+-----------
+If one term is `inj{s1, s2}(p1)` where p1 is an overloaded application, and the
+other is `p2`, an overload of `p1`, or if both terms are injections applied
+to overloaded constructors, or if one term is an overloaded symbol and the
+other is an injection applied to a variable, or if one term is an injection
+applied to an overloaded symbol and the other is an injection applied to a
+variable, we handle these cases via the overload simplifier, which returns one
+of four cases:
+
+* Clash: unification fails
+* Simple resolution: decompose the two terms returned by the simplifier.
+* Narrowing resolution: constrain the solution with the constraint returned by
+  the simplifier and decompose the terms returned.
+
+Otherwise if one is an injection and the other is an overloaded symbol, fail.
+
+Map
+---
+
+If both patterns are of a map sort, we first normalize each map by categorizing
+each subterm according to one of three categories:
+
+1. Elements
+2. Variables
+3. Functions (other than `.Map`, `_Map_`, and `|->`)
+
+This process may yield the value \bottom for either map, in which case, fail.
+
+First, we remove all subterms present in both maps from each map.
+
+We then consider 2 cases:
+
+1. Either map still contains functions. In this case, constrain with an
+   equality between the two maps.
+2. Both maps contain no functions. In this case, apply AC unification to the
+   two maps.
+
+### AC Unification 
+
+First, we consider the following special cases:
+
+* If both maps are empty, discharge.
+* If one map has 0 elements and 1 variable, treat it as a variable pattern
+  instead, and perform variable abstraction on the other pattern, then follow
+the procedure for AC binding.
+* If both maps have 1 element and 0 variables, decompose with the keys and
+  values.
+* Otherwise, if one map is empty and the other has 0 elements, bind each
+  variable of the other map to the empty map.
+* Otherwise, if one map is empty (and the other has elements), fail.
```

Comment:
I mean, we would have to pairwise compare the keys of the maps to do this, which would be a quadratic operation in the size of both maps. I don't think it would actually be faster.

Reply:
This falls into the same category as the problem above; what you suggest sounds like a useful optimization, but in the absence of a real world program taking a significant amount of execution time doing ac unification, we don't really want to tinker with a standard algorithm.

Reply:
I'm running the Multisig proofs on this PR's branch to see if there is any performance loss. Interestingly enough, around 50% of the Multisig proofs failed so far (I didn't finish running all the proofs). It could be that I'm using some unsupported failure, or that I am relying on a bug that got fixed by this PR, but I'd like to figure out what happens before approving.

Reply:
Point me to a specific proof that failed and I can take a look. It's possible that if you previously relied on built in unification for `in_keys`, but don't import `MAP-SYMBOLIC`, this could explain the failures. Or it may be a bug of some kind.

Reply:
Also posted on Slack:

To reproduce:

test.k:
```
module TEST
  imports BOOL
  imports INT
  imports MAP

  syntax Address ::= address(Int)
  syntax U ::= u(Int)
  syntax Maybe ::= just(KItem) | "nothing"

  syntax KItem ::= mapLookup(KItem, Map)

  rule mapLookup(K:KItem, ((K |-> V:KItem) _:Map) #as M:Map) => just(V)
    ensures K in_keys(M)
  rule mapLookup(K:KItem, M:Map) => nothing
    ensures notBool (K in_keys(M))
    [owise]
endmodule
```
proof-test.k:
```
module PROOF-TEST
  imports TEST

  claim (mapLookup(address(X), address(X) |-> u(_) M:Map) => ?_:Maybe) ~> M:Map ~> _:K
endmodule
```
command line:
```
kompile --backend haskell test.k && kprove proof-test.k
```